1.18.2 item launching expansion + Holding compatibility

Alice0Rose opened the pull request:
- Fixes Holding enchantment compatibility for latest version of CoFH Core
- Fixes item launching for Thermal grenades, TNT, and minecarts
- Fixes boat launching to use same boat type as item
- Adds item launching for Botania minecart
- Adds item launching for tridents
- Adds item launching for armor stands
- Adds item launching for micromissiles

Commit:
- Fixed Holding Enchantment integration to work with the newest version of CoFH Core
- Added IMC support (as part of getting Holding integration working)
- Started work on expanding item launching

Commit:
- Reimplemented Thermal Foundation TNT and Grenade launching, along with adding Thermal Locomotion Minecart launching
- Fixed boat launching to use the boat type of the item, and also to not push the player down when launched

Commit:
- Added Botania minecart launching
- Added Thermal charge launching (disabled until Thermal adds a default despawn time to the entities, so they don't last forever)
- Finished updating item launching

desht commented:
Thanks for this, looks pretty extensive.  I'll try & review it in the next few days.   Assuming all is well, are you interested in cherry-picking the changes into 1.19.2 as well?

desht commented on src/main/java/me/desht/pneumaticcraft/common/thirdparty/cofhcore/ThermalFoundationExplosiveLaunching.java:
```diff
@@ -41,6 +41,13 @@ public static void registerExplosiveLaunchBehaviour() {
         register("nuke_grenade");
 
+        // Adds thermal charges to launch map
+        // TODO: Find a way to get these to despawn after being launched, currently they float around forever
```
Are these coded like vanilla fireballs?  I did fix a similar problem with those a while back, in commit 7f02173b67fe06f03e3a57d79e87d73059fc9bd2

desht commented on src/main/java/me/desht/pneumaticcraft/common/network/PacketChestplateLauncher.java:
```diff
@@ -66,19 +70,31 @@ private void handleLaunch(ServerPlayer player) {
+            ItemStack toFire = stack;
+
+            // Split stack only for items that are consumed when dispensed (not micromissiles)
+            if (!(stack.getItem() == ModItems.MICROMISSILES.get())) {
```
Just `!=` here

desht commented on src/main/java/me/desht/pneumaticcraft/common/thirdparty/PneumaticcraftIMC.java:
```diff
@@ -0,0 +1,40 @@
+ * Class containing manager for sending all IMC messages to other mods
+ */
+@Mod.EventBusSubscriber(modid = Names.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
+public class PneumaticcraftIMC {
```
I presume this a message queue to collect all the messages we need to send at the right time?  If so, I'd suggest an `ArrayDeque` rather than `ArrayList`, just for tidiness (no need to keep queued messages around after they've been sent)

Alice0Rose replied:
I changed it, but out of curiosity, why is an `ArrayDeque` better than an `ArrayList` here?

desht replied:
I mean, an `ArrayList` is OK, but I like to fit the data structure to the task.  And the task here is a queue of message, so `ArrayDeque` is just a more natural fit.

desht replied:
Also means code like this can be used:
```java
while (!iMCMessageQueue.isEmpty()) {
    InterModComms.IMCMessage message = iMCMessageQueue.pop();
    InterModComms.sendTo(Names.MOD_ID, message.modId(), message.method(), message.messageSupplier());
}
```

And it occurred to me after that this can be called off-thread (from `FMLCommonSetupEvent`) so I made it a `ConcurrentLinkedQueue`.  A bit paranoid, but never a bad idea to be over-careful when it comes to threading.

desht commented on src/main/java/me/desht/pneumaticcraft/common/thirdparty/cofhcore/HoldingEnchantableProvider.java:
```diff
@@ -17,66 +17,41 @@
     }
 
-    public record COFHVolumeModifier(Enchantment holding) implements ItemVolumeModifier {
+    public record C0FHVolumeModifier(Enchantment holding) implements ItemVolumeModifier {
```
Any particular reason for the 0 (zero) instead of O (capital 'o') here?

Alice0Rose replied:
I honestly have no idea why I changed that, must have been a typo when I was retyping the name or something.

desht commented on gradle.properties:
```diff
@@ -37,6 +37,13 @@ botania_version=1.18.2-432
 cc_tweaked_version=1.100.5
 immersive_engineering_version=1.18.2-8.0.2-pre.12
 ctm_version=1.1.2.6
+cofh_core_curse_id = 4112120
```
Can you also update `mods.toml` to require the min version of CoFH which uses IMC for registering holding compat?

Alice0Rose replied:
I can, but that will make the mod unusable with CoFH Core until CoFH Core releases version 9.1.2 (which is the version the IMC compat is added in). They removed the old way of registering for Holding enchantments a few versions ago, and only have the IMC way as part of one of their latest versions which is only on their Github right now.

desht replied:
Ah, I thought that had already been released. OK, as long as it doesn't actually crash with the old version (i.e. just does nothing), that should be alright.

Commit:
- Fixes everything mentioned in review but the Thermal charge launching
- Dependency for CoFH Core is in the mods.toml, but commented out for now until they publicly release the latest version.
